src/functions/GroundMotionSim.C

```diff
@@ -0,0 +1,154 @@
+/*************************************************/
+/*           DO NOT MODIFY THIS HEADER           */
+/*                                               */
+/*                     MASTODON                  */
+/*                                               */
+/*    (c) 2015 Battelle Energy Alliance, LLC     */
+/*            ALL RIGHTS RESERVED                */
+/*                                               */
+/*   Prepared by Battelle Energy Alliance, LLC   */
+/*     With the U. S. Department of Energy       */
+/*                                               */
+/*     See COPYRIGHT for full restrictions       */
+/*************************************************/
+
+#include "GroundMotionSim.h"
+registerMooseObject("MastodonApp", GroundMotionSim);
+
+InputParameters
+GroundMotionSim::validParams()
+{
+  InputParameters params = Function::validParams();
+  params.addRequiredParam<Real>("M", "Magnitude of Event");
+  params.addRequiredParam<Real>("R", "Joyner-Boore or Epicentral Distance");
+  params.addRequiredParam<Real>("Vs30", "Shear Wave Velocity at 30m");
+  MooseEnum F("SS TF NF", "SS");
+  MooseEnum R_type("Rjb Repi", "Rjb");
+  params.addParam<MooseEnum>("F", F, "Style of Faulting");
+  params.addParam<MooseEnum>(
+      "R_type", R_type, "Indicates whether R distance is Joyner-Boore or Epicentral distance");
+  params.addParam<unsigned int>("n", 1012, "Random Seed");
+  params.addParam<Real>("scale_factor", 9.81, "acceleration scaling factor");
+  params.set<std::vector<Real>>("xy_data") = {};
+  params.addClassDescription("Calculates a ground acceleration history using a ground motion model "
+                             "written in a python script");
+  return params;
+}
+
+// Define arbitrarily large size of preallocated arrays for t and Acc
+#define BUFSIZE 20000
+
+GroundMotionSim::GroundMotionSim(const InputParameters & parameters)
+  : PiecewiseLinearBase(parameters), _scale_factor(getParam<Real>("scale_factor"))
+{
+
+  const auto M = this->template getParam<Real>("M");
+  const auto R = this->template getParam<Real>("R");
+  const auto V = this->template getParam<Real>("Vs30");
+  const auto n = this->template getParam<unsigned int>("n");
+  std::string R_type = this->template getParam<MooseEnum>("R_type");
+  std::string F = this->template getParam<MooseEnum>("F");
+
+  // Convert input parameters to strings
+  std::string Mstr = std::to_string(M);
+  std::string Rstr = std::to_string(R);
+  std::string Vstr = std::to_string(V);
+  std::string nstr = std::to_string(n);
+
+  // Get path to simulation script
+  std::string PATH = getDataFileNameByName("GMSim.py");
```

Comment:
What if the user has GMSim.py at multiple locations? How does this code pick which one to use then?

Reply:
The user should not have it in multiple locations. As part of this merge, a new folder called "data" is added in the top level folder. This is where GMSim.py is contained. It is not a user created script, it is the script that generates the actual ground motion.